airbyte-integrations/base-java/src/main/java/io/airbyte/integrations/javabase/DestinationConsumer.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.javabase;
+
+import io.airbyte.commons.functional.CheckedConsumer;
+import java.io.IOException;
+
+/**
+ * Lifecyle:
+ * * Consumer object is instantiated.
+ * * It received messages via {@link DestinationConsumer#accept(T)}
+ * * Upon receiving the last message (assuming no failures) {@link DestinationConsumer#complete()}
```

Comment:
A little on the fence on this one. One of the common patterns we are going to use is write data to some sort of tmp table and then if everything succeeds overwrite the existing table with the tmp table. With the complete method, I'm trying to make that pattern easy to do. Otherwise `close()` has to try to detect if writing to the tmp table was successful and whether or not to copy the data from tmp to the main table. That'll be messy to write and tricky to figure out since `close` doesn't take any args.

We can definitely make a default implementation of complete that is a no op.

Reply:
Can we assume that the consumer keeps the state of success/failure internally? We can always have an abstract class that has a more refined interfac.

Reply:
We can. It is adding responsibility / friction on an OSS contributor to figure this out though which I think runs counter to what we're trying to do.

Reply:
it becomes something like this, which i'm not a huge fan of. yeah, we could do an abstract class to make this less rough, but it seems ultimately the abstraction we're trying to retain by doing this is the one currently in the interface.

```
  public static class RecordConsumer implements DestinationConsumer<SingerMessage> {
    private boolean didFail = false;

    @Override
    public void accept(SingerMessage singerMessage) throws Exception {
      try {
      // write to tmp table
      } catch (Exception e) {
        didFail = true;
        throw e;
      }
    }

    @Override
    public void close() throws Exception {
      if(!didFail) {
        // copy tmp table to main table.
      }
      // close connection.
    }
  }
```

did that convince you at all? 😉

Reply:
It could have except that you change the contract on complete in a recent commit which makes it the same as close.